### PR TITLE
Add failing self referencing habtm spec

### PR DESCRIPTION
**Pull request by rmachielse**
Hi @estolfo 

This pull request adds a failing test case related to this issue:
https://jira.mongodb.org/browse/MONGOID-4273

### CODE REVIEW

**Comment by johnnyshields** [comment]
@estolfo we should get a fix here before 6.0

**Comment by rmachielse** [comment]
@estolfo @johnnyshields I have rebased this branch on the latest master, unfortunately this is still a problem. Do you think it is a valid problem? This pr has been open for a long time. Otherwise I will close it.

**Comment by johnnyshields** [comment]
@durran @estolfo please help with a fix

**Comment by johnnyshields** [comment]
@rmachielse can you check if https://github.com/mongodb/mongoid/pull/4383 fixes this issue?

**Comment by rmachielse** [comment]
@johnnyshields I have rebased my branch on those changes here and run the tests locally:

https://github.com/mongodb/mongoid/compare/master...feedbackfruits:spec/self-referencing-habtm-on-refactor

Unforturtonately it doesn't make a difference

File: spec/mongoid/relations/referenced/many_to_many_spec.rb [hunk 1]
```diff
@@ -814,6 +814,82 @@
     end
   end
 
+  describe "when self-referincing" do
```
Comment (johnnyshields) [inline spec/mongoid/relations/referenced/many_to_many_spec.rb]: typo: should be self-referencing

**Comment by johnnyshields** [comment]
@estolfo this issue is worthy of investigation... basically it seems that `reload` is clearing the reference value here... very strange

File: spec/mongoid/relations/referenced/many_to_many_spec.rb (hunk 1 above)
Comment (rmachielse) [inline spec/mongoid/relations/referenced/many_to_many_spec.rb]: fixed!

**Comment by rmachielse** [comment]
@johnnyshields well that means it's not persisted right?

**Comment by johnnyshields** [comment]
@rmachielse i guess so--did you inspect the values in MongoDB itself?

**Comment by rmachielse** [comment]
@johnnyshields yes, we had a workaround to persist the values again manually. However we are not using mongoid anymore now.

**Comment by johnnyshields** [comment]
So basically the issue here is that `update` of a self-referencing-habtm with the same value as it originally had seems to clear the value of its related partner.

One hunch is that we're doing an atomic `push` then a `pull`, which would set the same value then remove it.

**Comment by rmachielse** [comment]
@johnnyshields yes indeed.

Our workaround looked like this:

```ruby
after_save do
  parents.each do |parent|
    parent.add_to_set(child_ids: id)
  end
end
```

**Comment by estolfo** [comment]
Fixed in this commit: https://github.com/mongodb/mongoid/commit/8ccd33185158b53a79319a8d670907a744d7ecc9

**Pull request closed without merging**